### ISSUE TITLE
Faster joins: incoming `/send` requests block during a faster join

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#13684]
Things like typing updates can cause `/send` requests to block while there is a faster-join going on. This is bad because it means that all traffic ends up getting stacked up.

```
2022-08-31 22:22:35,106 - synapse.storage.util.partial_state_events_tracker - 169 - INFO - PUT-795 - Awaiting un-partial-stating of room !OGEhHVWSdvArJzumhm:matrix.org
Stack (most recent call last):
...
  File "/usr/local/lib/python3.9/site-packages/synapse/util/async_helpers.py", line 227, in _concurrently_execute_inner
    await maybe_awaitable(func(value))
  File "/usr/local/lib/python3.9/site-packages/synapse/federation/federation_server.py", line 522, in _process_edu
    await self.registry.on_edu(edu.edu_type, origin, edu.content)
  File "/usr/local/lib/python3.9/site-packages/synapse/federation/federation_server.py", line 1441, in on_edu
    await handler(origin, content)
  File "/usr/local/lib/python3.9/site-packages/synapse/handlers/typing.py", line 365, in _recv_edu
    domains = await self._storage_controllers.state.get_current_hosts_in_room(
  File "/usr/local/lib/python3.9/site-packages/synapse/storage/controllers/state.py", line 525, in get_current_hosts_in_room
    await self._partial_state_room_tracker.await_full_state(room_id)
  File "/usr/local/lib/python3.9/site-packages/synapse/logging/opentracing.py", line 876, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/usr/local/lib/python3.9/site-packages/synapse/storage/util/partial_state_events_tracker.py", line 169, in await_full_state
    logger.info(
```

(I think this might be fundamentally the same as #13288)

**Comment by MatMaul** [comment]
I am adding tests and fixing if needed the receiving of EDUs during fast joins:
- `m.typing` should be fixed by https://github.com/matrix-org/complement/pull/465
- `m.direct_to_device` is ok
- `m.presence` is not crashing the transaction however it doesn't seem to go down the sync afterwards, I haven't investigated yet
- `m.receipt` untested yet, TODO
- `m.signing_key_update` untested yet, TODO
- `m.device_list_update` untested yet, TODO

The PR for the complement tests is [here](https://github.com/matrix-org/complement/pull/465).

**Comment by richvdh** [comment]
fixed by #13830. Thanks @MatMaul !
